crates/services/txpool_v2/src/pool.rs

```diff
@@ -162,18 +198,95 @@ where
-        if self.storage.count() >= self.config.max_txs {
+    /// Check if the pool has enough space to store a transaction.
+    /// It will try to see if we can free some space depending on defined rules
+    /// If the pool is not full, it will return an empty list
+    /// If the pool is full, it will return the list of transactions that must be removed from the pool along all of their dependent subtree
+    /// If the pool is full and we can't make enough space by removing transactions, it will return an error
+    fn check_pool_size_available(
+        &self,
+        tx: &PoolTransaction,
+        collision: &Option<S::StorageIndex>,
+        dependencies: &[S::StorageIndex],
+    ) -> Result<Vec<S::StorageIndex>, Error> {
+        let tx_gas = tx.max_gas();
+        let bytes_size = tx.metered_bytes_size();
+        if self.current_gas.saturating_add(tx_gas) <= self.config.pool_limits.max_gas
+            && self.current_bytes_size.saturating_add(bytes_size)
+                <= self.config.pool_limits.max_bytes_size
+            && self.storage.count().saturating_add(1) <= self.config.pool_limits.max_txs
+        {
+            return Ok(vec![]);
+        }
+        // If the transaction has a collision verify that by removing the transaction we can free enough space
+        // otherwise return an error
+        if let Some(collision) = collision {
+            let collision_data = self.storage.get(collision)?;
+            let new_current_gas = self
+                .current_gas
+                .saturating_sub(collision_data.dependents_cumulative_gas);
+            let new_current_bytes_size = self
+                .current_bytes_size
+                .saturating_sub(collision_data.dependents_cumulative_bytes_size);
+            if new_current_gas < self.config.pool_limits.max_gas
+                && new_current_bytes_size < self.config.pool_limits.max_bytes_size
+                && self.storage.count().saturating_sub(1)
+                    < self.config.pool_limits.max_txs
+            {
+                return Ok(vec![*collision]);
+            } else {
+                return Err(Error::NotInsertedLimitHit);
+            }
+        }
+
+        // If the transaction has a dependency and the pool is full, we refuse it
+        if !dependencies.is_empty() {
             return Err(Error::NotInsertedLimitHit);
         }
-        Ok(())
+
+        // Here the transaction has no dependencies and no collision which means that it's an executable transaction
+        // and we want to make space for it
+        let mut removed_transactions = vec![];
+        let mut gas_left = self.current_gas.saturating_add(tx_gas);
+        let mut bytes_left = self.current_bytes_size.saturating_add(bytes_size);
+        let mut txs_left = self.storage.count().saturating_add(1);
+        let current_ratio = Ratio::new(tx.tip(), tx_gas);
+        let mut sorted_txs = self
+            .storage
+            .get_worst_ratio_tip_gas_subtree_roots()?
```

Comment:
The problem is that the sorting will be redo each time we delete a child of someone as the `cumulative_gas`/`cumulative_tip` of a lot of transactions will change. So I was thinking that computing it during insert could be best but I will change.
But I don't really know how this can be done outside of the storage has it's the storage that is in charge of modifying the `cumulative_gas`/`cumulative_tip` and the rest of the txpool don't know when these values are updated

Reply:
Following our discussion, I changed this, now it uses a function inside the selection algorithm which seems to be the most appropriate for this. I have also fixed a bug that was there